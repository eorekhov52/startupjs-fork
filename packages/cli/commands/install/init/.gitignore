# <startupjs>
# local SQLite database which is used by default instead of an actual MongoDB
/local.db
# </startupjs>
