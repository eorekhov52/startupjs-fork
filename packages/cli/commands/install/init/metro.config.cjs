const { getDefaultConfig } = require('startupjs/metro-config')

const config = getDefaultConfig(__dirname)

module.exports = config
