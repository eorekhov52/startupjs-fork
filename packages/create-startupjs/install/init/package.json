{
  "type": "module",
  "dependencies": {
    "startupjs": "^%%STARTUPJS_MINOR_VERSION%%"
  },
  "scripts": {
    "postinstall": "npx startupjs postinstall",
    "build": "npx startupjs build",
    "start-production": "npx startupjs start-production"
  }
}
