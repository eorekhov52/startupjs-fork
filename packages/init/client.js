export { default } from './lib/native/index.js'
