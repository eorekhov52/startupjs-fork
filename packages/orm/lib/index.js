import { singletonMemoize } from '@startupjs/cache'
import racer from 'racer'
import promisifyRacer from './promisifyRacer.js'

const Model = racer.Model

global.STARTUP_JS_ORM = {}

export default function (racer) {
  if (racer.orm) return

  racer._orm = global.STARTUP_JS_ORM
  racer.orm = function (pattern, OrmEntity, alias) {
    const name = alias || pattern
    if (global.STARTUP_JS_ORM[name]) throw alreadyDefinedError(pattern, alias)

    // NOTE
    // if same OrmEntity will be passed for different collections
    // then they will all have the same collection name
    if (!OrmEntity.collection) {
      const match = pattern.match(/^[^.]+/)
      if (match) OrmEntity.collection = match[0]
    }

    global.STARTUP_JS_ORM[name] = {
      pattern,
      regexp: patternToRegExp(pattern),
      OrmEntity
    }
  }

  Model.prototype.at = function (subpath, alias) {
    const path = this.path(subpath)
    return this.scope(path, alias)
  }

  Model.prototype._scope = function (path) {
    const ChildModel = Model.ChildModel
    const model = new ChildModel(this)
    model._at = path
    return model
  }

  // TODO: do NOT cache within built-in use* hooks from react-sharedb
  //       because they already handle caching internally
  Model.prototype.scope = singletonMemoize(function (path, alias) {
    if (alias) {
      if (global.STARTUP_JS_ORM[alias]) {
        return this.__createScopedModel(path, global.STARTUP_JS_ORM[alias].OrmEntity)
      } else {
        throw new Error(
          'Non-existent alias of the OrmEntity specified: ' +
            alias +
            '\n\n' +
            'Most likely you have specified the path incorrectly in ' +
            '".scope()" or ".at()"\n\n' +
            'The path must be passed as a single string separated by dots, ' +
            'for example:\n\n' +
            'CORRECT:\n' +
            "$root.at('users.' + userId)\n\n" +
            'INCORRECT:\n' +
            "$root.at('users', userId)"
        )
      }
    }

    const segments = this._dereference(this.__splitPath(path), true)
    const fullPath = segments.join('.')

    for (const name in global.STARTUP_JS_ORM) {
      const regexp = global.STARTUP_JS_ORM[name].regexp
      if (regexp.test(fullPath)) {
        return this.__createScopedModel(path, global.STARTUP_JS_ORM[name].OrmEntity)
      }
    }

    return this._scope(path)
  }, {
    cacheName: 'model',
    nestedThis: true
  })

  Model.prototype.__createScopedModel = function (path, OrmEntity) {
    let model
    if (OrmEntity.factory || OrmEntity.prototype.factory) {
      model = OrmEntity(this._scope(path), this)
      // if factory didn't return anything, return a simple scoped model
      if (!model) model = this._scope(path)
    } else {
      model = new OrmEntity(this)
    }
    model._at = path
    return model
  }

  Model.prototype.__splitPath = function (path) {
    return (path && path.split('.')) || []
  }

  promisifyRacer()
}

function patternToRegExp (pattern) {
  pattern = pattern.replace(/\$/g, '\\$').replace(/\./g, '\\.').replace(/\*/g, '([^\\.]*)')
  return new RegExp('^' + pattern + '$')
}

function alreadyDefinedError (pattern, alias) {
  let msg
  if (alias) {
    msg =
      "ORM entity with the alias '" +
      alias +
      "' is already defined. " +
      'Aliases must be unique. If you did already define the same ORM entity with ' +
      "that alias name, just don't specify the alias at all -- path pattern is sufficient."
  } else {
    msg = "ORM entity matching the same path pattern '" + pattern + "' is already defined."
  }
  return new Error(msg)
}

export const ChildModel = Model.ChildModel

function BaseModel () {
  Model.ChildModel.apply(this, arguments)
}

BaseModel.prototype = Object.create(Model.ChildModel.prototype)
BaseModel.prototype.constructor = BaseModel

BaseModel.prototype.getId = function () {
  const actualField = this.dereferenceSelf()
  return actualField.leaf()
}

BaseModel.prototype.getCollection = function () {
  let collection = this.constructor.collection

  // fallback when orm is factory
  if (!collection) {
    const model = this.root
    const actualField = this.dereferenceSelf()
    collection = model._splitPath(actualField.path())[0]
  }

  return collection
}

BaseModel.prototype.dereferenceSelf = function () {
  const model = this.root
  const segments = model._splitPath(this.path())
  return model.scope(model._dereference(segments, true).join('.'))
}

BaseModel.associations = []

BaseModel.addAssociation = function (association) {
  this.associations = this.associations.concat(association)
}

BaseModel.prototype.getAssociations = function () {
  return this.constructor.associations
}

export { BaseModel }
export * from './associations/index.js'

/**
 * Pick properties from json-schema to be used in a form.
 * Supports simplified schema (just the properties object) and full schema.
 * Performs extra transformations like auto-generating `label`.
 * `createdAt`, `updatedAt`, `id`, `_id` fields are excluded by default.
 * @param {Object} schema
 * @param {Object|Array} options - exclude or include fields. If array, it's the same as passing { include: [...] }
 * @param {Array} options.include - list of fields to pick (default: all)
 * @param {Array} options.exclude - list of fields to exclude (default: none)
 */
export function pickFormFields (schema, options) {
  let include, exclude
  if (Array.isArray(options)) {
    include = options
  } else {
    ;({ include, exclude } = options || {})
  }
  exclude ??= []
  if (!schema) throw Error('pickFormFields: schema is required')
  schema = JSON.parse(JSON.stringify(schema))
  if (schema.type === 'object') {
    schema = schema.properties
  }
  for (const key in schema) {
    if (
      include?.includes(key) ||
      (!include && !exclude.includes(key) && !DEFAULT_EXCLUDE_FORM_FIELDS.includes(key))
    ) {
      const field = schema[key]
      if (!field.label) field.label = camelCaseToLabel(key)
    } else {
      delete schema[key]
    }
  }
  return schema
}

const DEFAULT_EXCLUDE_FORM_FIELDS = ['id', '_id', 'createdAt', 'updatedAt']

// split into words, capitalize first word, make others lowercase
function camelCaseToLabel (str) {
  return str
    .replace(/([A-Z])/g, ' $1')
    .toLowerCase()
    .replace(/^./, (s) => s.toUpperCase())
}
