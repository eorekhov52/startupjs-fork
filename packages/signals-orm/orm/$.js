// this is just the $() function implementation.
// The actual $ exported from this package is a Proxy targeting the dataTree root,
// and this function is an implementation of the `apply` handler for that Proxy.
import { observe, unobserve } from '@nx-js/observer-util'
import getSignal from './getSignal.js'
import { set as _set, del as _del } from './dataTree.js'

export const LOCAL = '$local'
const DELETION_DELAY = 0
let counter = 0

const valuesFr = new FinalizationRegistry(id => {
  _del([LOCAL, id])
})

const reactionsFr = new FinalizationRegistry(([reaction, id]) => {
  unobserve(reaction)
  // don't delete data right away to prevent dependent reactions which are also going to be GC'ed
  // from triggering unnecessarily
  setTimeout(() => _del([LOCAL, id]), DELETION_DELAY)
})

export default function $ (value) {
  if (typeof value === 'function') {
    return reaction$(value)
  } else {
    return value$(value)
  }
}

function value$ (value) {
  const id = `_${counter}`
  counter += 1
  _set([LOCAL, id], value)
  const $value = getSignal([LOCAL, id])
  valuesFr.register($value, id)
  return $value
}

function reaction$ (fn) {
  const id = `_${counter}`
  counter += 1
  const reactionScheduler = reaction => runReaction(reaction, id)
  const reaction = observe(fn, { lazy: true, scheduler: reactionScheduler })
  runReaction(reaction, id)
  const $value = getSignal([LOCAL, id])
  reactionsFr.register($value, [reaction, id])
  return $value
}

function runReaction (reaction, id) {
  const newValue = reaction()
  _set([LOCAL, id], newValue)
}
