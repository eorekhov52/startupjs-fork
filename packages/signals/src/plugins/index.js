import * as queryMethods from './queryMethods'
import * as queryMap from './queryMap'
import * as friendlyNames from './friendlyNames'
import * as queryData from './queryData'
import * as _default from './_default'

// IMPORTANT! Order matters here since plugins will be applied in this order!
export default [
  queryMethods,
  queryMap,
  friendlyNames,
  queryData,
  // IMPORTANT! _default must always be last:
  _default
].filter(Boolean)
