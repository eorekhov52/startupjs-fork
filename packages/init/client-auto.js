import init from './client.js'

// TODO: Check if we have server enabled. For this get config value from @startupjs/config
init()

export default () => {}
