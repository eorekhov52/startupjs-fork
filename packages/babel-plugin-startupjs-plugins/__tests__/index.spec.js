const pluginTester = require('babel-plugin-tester').default
const plugin = require('../index')
const { name: pluginName } = require('../package.json')

pluginTester({
  plugin,
  pluginName,
  snapshot: true,
  pluginOptions: {
    env: 'server',
    root: process.cwd() + '/fixtures'
  },
  tests: {
    'Ignores files without a magic import': /* js */`
      import something from 'startupjs/something'
    `,
    'Processes files with a magic import': /* js */`
      import plugins from 'startupjs/plugins'
      import init from 'startupjs/init'
      import orm from '../model'

      init({ orm, plugins })
    `
  }
})
