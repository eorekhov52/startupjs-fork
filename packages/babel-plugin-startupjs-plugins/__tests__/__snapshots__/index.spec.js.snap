// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@startupjs/babel-plugin-startupjs-plugins Ignores files without a magic import: Ignores files without a magic import 1`] = `

import config from './startupjs.config.js'
console.log(config)

      ↓ ↓ ↓ ↓ ↓ ↓

import config from "./startupjs.config.js";
console.log(config);


`;

exports[`@startupjs/babel-plugin-startupjs-plugins Processes files with a magic import: Processes files with a magic import 1`] = `

import { registry } from 'startupjs/registry'
import config from './startupjs.config.magic.js'
import dummy from '@dummy/dummy'

registry.init(config)

const x = 'xxx'
dummy(x)

export default () => {}

      ↓ ↓ ↓ ↓ ↓ ↓

import { registry } from "startupjs/registry";
import config from "../../startupjs.config.js";
import dummy from "@dummy/dummy";
import _default from "module-1/plugin";
import _default2 from "module-1-plugin/thePlugin.plugin";
import _default3 from "../../dummyPlugin.plugin.js";
(() => {})([_default, _default2, _default3]);
registry.init(config);
const x = "xxx";
dummy(x);
export default () => {};


`;

exports[`@startupjs/babel-plugin-startupjs-plugins Test sample file from fixtures which loads config: Test sample file from fixtures which loads config 1`] = `

import { registry } from 'startupjs/registry'
import config from './startupjs.config.magic.js'

registry.init(config)

      ↓ ↓ ↓ ↓ ↓ ↓

import { registry } from "startupjs/registry";
import config from "../../startupjs.config.js";
import _default from "module-1/plugin";
import _default2 from "module-1-plugin/thePlugin.plugin";
import _default3 from "../../dummyPlugin.plugin.js";
(() => {})([_default, _default2, _default3]);
registry.init(config);


`;
